blog/2023-05-02-understanding-context-switching-and-its-impact-on-system-performance/index.md

```diff
@@ -0,0 +1,52 @@
+
+In addition, monitoring the number of processes in the run queue (Runnable dimension in `system.processes_state chart`), the number of TASKLET and SCHED softirqs, as well as the time spent handling these softirqs, you can identify if there is a high level of task scheduling activity on the system.
+
+In summary, detecting thrashing can be challenging, but monitoring the system's performance metrics, the number of context switches per second, the number of processes in the run queue, and the average time spent in the scheduler can help to identify its presence. By optimizing the performance of the processes or threads that are causing the most context switches, you may be able to alleviate the thrashing and improve system performance.
+
+## How to reduce the impact of context switching?
+
+Here are some options that can help reduce the impact of context switching on system performance:
+
+- **Increase CPU and memory resources:** More CPU and memory resources can reduce the frequency of context switching and provide more room for the system to handle multiple tasks simultaneously.
+
+- **Use process scheduling policies:** Operating systems provide various process scheduling policies to optimize CPU utilization and reduce context switching overhead. For example, the CFS (Completely Fair Scheduler) policy in Linux is designed to minimize context switching overhead.
+
+- **Use lightweight threads:** If you are developing the application that is causing extensive context switches, use lightweight threads, also known as user-level threads, which are managed by user-level code instead of the kernel. They are faster to create and switch between than kernel-level threads, which can reduce context switching overhead.
+
+- **Avoid unnecessary context switches:** Again for developers: context switching can be triggered unnecessarily by interrupting processes or threads that are waiting for I/O. You can avoid this by using non-blocking I/O operations or by using asynchronous I/O operations.
+
+- **Use CPU affinity:** [CPU affinity](https://en.wikipedia.org/wiki/Processor_affinity) is a technique that assigns a specific CPU to a process or thread. This can reduce the frequency of context switching and improve cache utilization.
+
+- **Use NUMA-aware scheduling:** [NUMA (Non-Uniform Memory Access) architectures](https://en.wikipedia.org/wiki/Non-uniform_memory_access) have multiple memory nodes, and scheduling processes and threads to run on the same memory node as their data can reduce the frequency of remote memory accesses and improve performance.
+
+- **Use real-time scheduling:** Real-time scheduling policies can give priority to time-critical processes, reducing context switching and ensuring that the most important tasks are completed on time.
+
+It's worth noting that some of these options may not be appropriate for all systems and applications, and the best approach will depend on the specific requirements and constraints of your system.
```

Comment:
Another sentence that says nothing. Yes, obviously. If we need a disclaimer like thjis, I'd put it in the sentence before the list.